File: Day-10/MobileKeypad.js
function countSequences(n) {
  // Keypad layout
  const keypad = [
      ['1', '2', '3'],
      ['4', '5', '6'],
      ['7', '8', '9'],
      ['*', '0', '#']
  ];

  // Directions: stay, up, down, left, right
  const directions = [[0, 0], [-1, 0], [1, 0], [0, -1], [0, 1]];

  // Initialize DP table: dp[i][j][length]
  const dp = Array.from({ length: 4 }, () =>
      Array.from({ length: 3 }, () =>
          Array(n + 1).fill(0)
      )
  );

  // Base case: sequences of length 1
  for (let i = 0; i < 4; i++) {
      for (let j = 0; j < 3; j++) {
          if (keypad[i][j] !== '*' && keypad[i][j] !== '#') {
              dp[i][j][1] = 1;
          }
      }
  }

  // Fill DP table for lengths from 2 to n
  for (let len = 2; len <= n; len++) {
      for (let i = 0; i < 4; i++) {
          for (let j = 0; j < 3; j++) {
              if (keypad[i][j] === '*' || keypad[i][j] === '#') continue;

              for (const [di, dj] of directions) {
                  const ni = i + di;
                  const nj = j + dj;

                  if (
                      ni >= 0 && ni < 4 &&
                      nj >= 0 && nj < 3 &&
                      keypad[ni][nj] !== '*' &&
                      keypad[ni][nj] !== '#'
                  ) {
                      dp[i][j][len] += dp[ni][nj][len - 1];
                  }
              }
          }
      }
  }

  // Sum all sequences of length n
  let total = 0;
  for (let i = 0; i < 4; i++) {
      for (let j = 0; j < 3; j++) {
          if (keypad[i][j] !== '*' && keypad[i][j] !== '#') {
              total += dp[i][j][n];
          }
      }
  }

  return total;
}

// Example usage
console.log("Number of sequences for n = 2:", countSequences(2)); // Output: 36
console.log("Number of sequences for n = 1:", countSequences(1)); // Output: 10
console.log("Number of sequences for n = 3:", countSequences(3)); // Output: 138

File: Day-10/MobileKeypad.py
def count_sequences(n):
    # Keypad layout
    keypad = [
        ['1', '2', '3'],
        ['4', '5', '6'],
        ['7', '8', '9'],
        ['*', '0', '#']
    ]

    # Directions: stay, up, down, left, right
    directions = [(0, 0), (-1, 0), (1, 0), (0, -1), (0, 1)]

    # Initialize DP table
    dp = [[[0] * (n + 1) for _ in range(3)] for _ in range(4)]

    # Base case: sequences of length 1
    for i in range(4):
        for j in range(3):
            if keypad[i][j] not in ('*', '#'):
                dp[i][j][1] = 1

    # Fill DP table
    for length in range(2, n + 1):
        for i in range(4):
            for j in range(3):
                if keypad[i][j] in ('*', '#'):
                    continue
                # Try all directions
                for di, dj in directions:
                    ni, nj = i + di, j + dj
                    if 0 <= ni < 4 and 0 <= nj < 3 and keypad[ni][nj] not in ('*', '#'):
                        dp[i][j][length] += dp[ni][nj][length - 1]

    # Sum all sequences of length n
    total = 0
    for i in range(4):
        for j in range(3):
            if keypad[i][j] not in ('*', '#'):
                total += dp[i][j][n]

    return total

# Example usage
print("Number of sequences for n = 2:", count_sequences(2))  # Output: 36